src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -0,0 +1,688 @@
+
+            private static DiagnosticDescriptor TypeNotSupported { get; } = new DiagnosticDescriptor(
+                id: "SYSLIB1030",
+                title: new LocalizableResourceString(nameof(SR.TypeNotSupportedTitle), SR.ResourceManager, typeof(FxResources.System.Text.Json.SourceGeneration.SR)),
+                messageFormat: new LocalizableResourceString(nameof(SR.TypeNotSupportedMessageFormat), SR.ResourceManager, typeof(FxResources.System.Text.Json.SourceGeneration.SR)),
+                category: SystemTextJsonSourceGenerationName,
+                defaultSeverity: DiagnosticSeverity.Warning,
+                isEnabledByDefault: true);
+
+            private static DiagnosticDescriptor DuplicateTypeName { get; } = new DiagnosticDescriptor(
+                id: "SYSLIB1031",
+                title: new LocalizableResourceString(nameof(SR.DuplicateTypeNameTitle), SR.ResourceManager, typeof(FxResources.System.Text.Json.SourceGeneration.SR)),
+                messageFormat: new LocalizableResourceString(nameof(SR.DuplicateTypeNameMessageFormat), SR.ResourceManager, typeof(FxResources.System.Text.Json.SourceGeneration.SR)),
+                category: SystemTextJsonSourceGenerationName,
+                defaultSeverity: DiagnosticSeverity.Warning,
+                isEnabledByDefault: true);
+
+            private readonly string _generationNamespace;
+
+            // TODO: consider public option for this.
+            // Converter-honoring logic generation can be simplified
+            // if we don't plan to have a feature around this.
+            private readonly bool _honorRuntimeProvidedCustomConverters = true;
+
+            private readonly GeneratorExecutionContext _executionContext;
+
+            /// <summary>
+            /// Types that we have initiated serialization metadata generation for. A type may be discoverable in the object graph,
+            /// but not reachable for serialization (e.g. it is [JsonIgnore]'d); thus we maintain a separate cache.
+            /// </summary>
+            private readonly HashSet<TypeMetadata> _typesWithMetadataGenerated = new();
+
+            /// <summary>
+            /// Types that were specified with System.Text.Json.Serialization.JsonSerializableAttribute.
+            /// </summary>
+            private readonly Dictionary<string, TypeMetadata>? _rootSerializableTypes;
+
+            public Emitter(in GeneratorExecutionContext executionContext, Dictionary<string, TypeMetadata>? rootSerializableTypes)
+            {
+                _executionContext = executionContext;
+                _generationNamespace = $"{executionContext.Compilation.AssemblyName}.JsonSourceGeneration";
+                _rootSerializableTypes = rootSerializableTypes;
+            }
+
+            public void Emit()
+            {
+                if (_rootSerializableTypes == null || _rootSerializableTypes.Count == 0)
```

Comment:
(minor) I'd move this check outside of the Emitter class to where it is called. It makes the logic there more straight forward:

1. Call the Parser
2. If the parser returned anything
    a. Call the Emitter